File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.c
struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {
    // If either list is empty, return the other
    if (!list1) return list2;
    if (!list2) return list1;

    // Decide the head of the merged list
    struct ListNode* head = NULL;
    if (list1->val < list2->val) {
        head = list1;
        list1 = list1->next;
    } else {
        head = list2;
        list2 = list2->next;
    }

    // `tail` will build the rest of the list
    struct ListNode* tail = head;

    while (list1 && list2) {
        if (list1->val < list2->val) {
            tail->next = list1;
            list1 = list1->next;
        } else {
            tail->next = list2;
            list2 = list2->next;
        }
        tail = tail->next;
    }

    // Attach the remaining part
    tail->next = list1 ? list1 : list2;

    return head;
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        l1=[]
        l2=[]
        n=list1
        while n:
            l1.append(n.val)
            n=n.next        
        n=list2
        while n:
            l2.append(n.val)
            n=n.next        
        l=sorted(l1+l2)
        new=None
        tail=None
        for i in l:
            n=ListNode(i)
            if new is None:
                new=n
                tail=n
            else:
                tail.next=n
                tail=n
        return new       